File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1) return INT_MAX;
        if (dividend == INT_MIN && divisor == 1) return INT_MIN;
        long long a = labs((long long)dividend);
        long long b = labs((long long)divisor);
        long long res = 0;
        while (a >= b) {
            long long temp = b, multiple = 1;
            while ((temp << 1) <= a) {
                temp <<= 1;
                multiple <<= 1;
            }
            a -= temp;
            res += multiple;
        }
        return ((dividend > 0) ^ (divisor > 0)) ? -res : res;
    }
};

File: sol/solution/0001-0100/0029. Divide Two Integers/Solution.kt
class Solution {
    fun divide(dividend: Int, divisor: Int): Int {
        if (dividend == Int.MIN_VALUE && divisor == -1) return Int.MAX_VALUE
        var a = dividend.toLong()
        var b = divisor.toLong()
        val negative = (a < 0) xor (b < 0)
        a = kotlin.math.abs(a)
        b = kotlin.math.abs(b)
        var result = 0L
        for (i in 31 downTo 0) {
            if ((a shr i) >= b) {
                result += 1L shl i
                a -= b shl i
            }
        }
        if (negative) result = -result
        if (result > Int.MAX_VALUE) return Int.MAX_VALUE
        if (result < Int.MIN_VALUE) return Int.MIN_VALUE
        return result.toInt()
    }
}